### PR TITLE
LibWeb: Implement window.length

### PR DESCRIPTION
**Pull request by networkException**
After seeing people have fun with the web on [twitter](https://twitter.com/mjackson/status/1565446175498440704) I thought we could also partake in the fun and gain a bit of spec compliance in the process :^)

### CODE REVIEW

File: Userland/Libraries/LibWeb/Bindings/WindowObject.cpp [hunk 1]
```diff
@@ -402,6 +403,20 @@ JS_DEFINE_NATIVE_FUNCTION(WindowObject::btoa)
...
+{
+    auto* impl = TRY(impl_from(vm));
+
+    // 1. If W's browsing context is null, then return 0.
+    auto* this_browsing_context = impl->associated_document().browsing_context();
+    if (!this_browsing_context)
+        return JS::Value(0);
+
+    // 2. Return the number of document-tree child browsing contexts of W's browsing context.
+    return JS::Value(this_browsing_context->document_tree_child_browsing_context_count());
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Bindings/WindowObject.cpp:417]: Neat! Please package this in a standalone function (https://html.spec.whatwg.org/multipage/window-object.html#number-of-document-tree-child-browsing-contexts) and consider fixing these in a second commit:

- https://github.com/SerenityOS/serenity/blob/987927a596cb773aee0069666d650e3f26f9f3ac/Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp#L77
- https://github.com/SerenityOS/serenity/blob/987927a596cb773aee0069666d650e3f26f9f3ac/Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp#L236

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 2]
```diff
@@ -745,6 +745,24 @@ BrowsingContext* BrowsingContext::choose_a_browsing_context(StringView name, boo
     return chosen;
 }
 
+// https://html.spec.whatwg.org/multipage/browsers.html#document-tree-child-browsing-context
+size_t BrowsingContext::document_tree_child_browsing_context_count() const
+{
+    size_t count = 0;
+    for_each_child_of_type<BrowsingContext>([this, &count](BrowsingContext const& child) {
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:752]:
```suggestion

    // A browsing context child is a document-tree child browsing context of parent if child is a child browsing context and child's container is in a document tree.
    for_each_child_of_type<BrowsingContext>([this, &count](BrowsingContext const& child) {
```

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 3]
```diff
@@ -745,6 +745,24 @@ BrowsingContext* BrowsingContext::choose_a_browsing_context(StringView name, boo
...
+            ++count;
+    });
+
+    return count;
+}
+
+// https://html.spec.whatwg.org/multipage/browsers.html#child-browsing-context
+bool BrowsingContext::is_child_of(BrowsingContext const& parent) const
+{
+    return container_document() && container_document()->browsing_context() == &parent;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:763]:
```suggestion
    // A browsing context child is said to be a child browsing context of another browsing context parent, if child's container document is non-null and child's container document's browsing context is parent.
    return container_document() && container_document()->browsing_context() == &parent;
```

File: Userland/Libraries/LibWeb/Bindings/WindowObject.cpp (hunk 1 above)
Comment (networkException) [reply Userland/Libraries/LibWeb/Bindings/WindowObject.cpp]: Having to call `impl_from` just to get to an instance of `WindowObject` is a bit weird, is there a better way?

File: Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp [hunk 4]
```diff
@@ -74,8 +74,8 @@ JS::ThrowCompletionOr<Optional<JS::PropertyDescriptor>> WindowProxy::internal_ge
         // 1. Let index be ! ToUint32(P).
         auto index = property_key.as_number();
 
-        // FIXME: 2. Let maxProperties be the number of document-tree child browsing contexts of W.
-        size_t max_properties = 0;
+        // 2. Let maxProperties be the number of document-tree child browsing contexts of W.
+        size_t max_properties = TRY(m_window->document_tree_child_browsing_context_count());
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp:78]: `auto`, same blow

File: Userland/Libraries/LibWeb/Bindings/WindowObject.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibWeb/Bindings/WindowObject.cpp]: That's how it currently works, but this will all change very soon when the wrappers go away.

**Comment by networkException** [comment]
I've changed the call to `for_each_child_of_type<BrowsingContext>` to `for_each_child` as there were issues with the lagom build. See https://discord.com/channels/830522505605283862/830739873119207426/1016836457027092490

File: Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp [hunk 5]
```diff
@@ -72,8 +72,8 @@ JS::ThrowCompletionOr<Optional<JS::PropertyDescriptor>> WindowProxy::internal_ge
         // 1. Let index be ! ToUint32(P).
         auto index = property_key.as_number();
 
-        // FIXME: 2. Let maxProperties be the number of document-tree child browsing contexts of W.
-        size_t max_properties = 0;
+        // Let maxProperties be the number of document-tree child browsing contexts of W.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Bindings/WindowProxy.cpp:75]: `// 2.`

**Review by linusg — APPROVED** [review]
Thanks, I hate it ;^)

**Pull request merged**